packages/core/src/browser/preferences/preference-service.ts

```diff
@@ -100,119 +124,132 @@ export class PreferenceServiceImpl implements PreferenceService, FrontendApplica
-            this.providerProvider(PreferenceScope.Workspace)
-        ];
-    }
-
-    protected reconcilePreferences(): void {
-        const changes: PreferenceChanges = {};
-        const deleted = new Set(Object.keys(this.preferences));
-        const preferences = this.parsePreferences();
-        // tslint:disable-next-line:forin
-        for (const preferenceName in preferences) {
-            deleted.delete(preferenceName);
-            const oldValue = this.preferences[preferenceName];
-            const newValue = preferences[preferenceName];
-            if (oldValue !== undefined) {
-                if (!JSONExt.deepEqual(oldValue, newValue)) {
-                    changes[preferenceName] = { preferenceName, newValue, oldValue };
-                    this.preferences[preferenceName] = deepFreeze(newValue);
+
+    protected createProviders(): void {
+        PreferenceScope.getScopes().forEach(scope =>
+            this.doCreateProvider(scope)
+        );
+    }
+
+    protected reconcilePreferences(changes: PreferenceChanges): void {
+        const changesToEmit: PreferenceChanges = {};
+        for (const prefName of Object.keys(changes)) {
+            const change = changes[prefName];
+            if (this.schema.isValidInScope(prefName, PreferenceScope.FOLDER)) {
+                changesToEmit[prefName] = change;
+                this.onPreferenceChangedEmitter.fire(change);
+                continue;
+            }
+            for (const s of PreferenceScope.getReversedScopes()) {
+                if (this.schema.isValidInScope(prefName, s)) {
+                    const p = this.providersMap.get(s);
+                    if (p) {
+                        const value = p.get(prefName);
+                        if (s > change.scope && value !== undefined && value !== null) {
+                            // preference defined in a more specific scope
+                            break;
+                        } else if (s === change.scope) {
+                            changesToEmit[prefName] = change;
+                            this.onPreferenceChangedEmitter.fire(change);
+                        }
+                    }
                 }
-            } else {
-                changes[preferenceName] = { preferenceName, newValue };
-                this.preferences[preferenceName] = deepFreeze(newValue);
             }
         }
-        for (const preferenceName of deleted) {
-            const oldValue = this.preferences[preferenceName];
-            changes[preferenceName] = { preferenceName, oldValue };
-            this.preferences[preferenceName] = undefined;
-        }
-        this.onPreferencesChangedEmitter.fire(changes);
-        // tslint:disable-next-line:forin
-        for (const preferenceName in changes) {
-            this.onPreferenceChangedEmitter.fire(changes[preferenceName]);
+        if (Object.keys(changesToEmit).length > 0) {
+            this.onPreferencesChangedEmitter.fire(changesToEmit);
```

Comment:
before it would compute proper composite change event, filter out irrelevant changes, like not it just dump raw changes

Reply:
it does compute proper composite change event, the logic is between lines 155-174